### ISSUE TITLE
Reporting Violations of Community Guidelines (Part 2)

### ISSUE THREAD

**Issue by crookedneighbor** [HabitRPG/habitica#4403]
First part: #4339 
## Features
- [x] Allow admins to flag and un-flag messages without refreshing
- [x] Allow users to flag system messages that display users with inappropriate usernames. Customize report modal to reflect this.
- [ ] Add report dialog to block user button - https://github.com/HabitRPG/habitica/issues/9486
- [x] Allow reporters to add a message to the mods when reporting abuse
## Discussion
- [ ] If a user has more than X flags on any of their posts in a certain time period, all future posts from them are automatically hidden.
- [ ] Automatic email sent to the person who wrote the offending message when post is hidden explaining that the community believed the post was in violation of the community guidelines, assuring the person that in all likelihood, their error was done in ignorance and encouraging them to review the community guidelines to avoid future posts from being hidden.
## Part 3 Ideas
- [ ] List of reported users that only admins can see
- [ ] Mods can assign comments (similar to contributor tier notes) and scores to infractions (1-5), after a user's total score reaches a certain threshold, chat privileges are revoked

_Edit: The following items were edited above for clarity_
- Some kind of rating system for severity of abuse -> Mods can assign comments (similar to contributor tier notes) and scores to infractions (1-5), after a user's total score reaches a certain threshold, chat privileges are revoked
##

**Comment by crookedneighbor** [comment]
@Alys and @lemoness I tried to pull the relevant information from the G+ discussion. Feel free to edit as needed.

**Comment by Alys** [comment]
I love all of this except maybe this one: **"Some kind of rating system for severity of abuse"**. Would that be for the users to use? I think having to chose a severity rating would be a bit stressful (I know I'd worry about getting it "right") and it would probably produce inconsistent results since some users would have different ideas about what's bad. I'm not sure it would help us much either; I think we're likely to look at the emails as soon as we see them and having some flagged as high/low priority wouldn't change that.

**Comment by crookedneighbor** [comment]
I pulled that out of this from Mel P.:

> In future we should also attempt a grading system for infraction severity. I'm suggesting a 5 point system with some broad guidelines as to what fits in each score and then a total score to determine whether a user is banned from chat and for how long. 
> That way we ensure consistency and also help future mods continue our work. ﻿

My abbreviated version missed much of the nuance of the suggestion. I'll edit the checklist to more accurately represent the idea.

**Comment by crookedneighbor** [comment]
Allowing reporters to add messages to reports of abuse and revoking chat privileges after x number of flags seem pretty simple to do.

I think compiling a list of users with flags and commenting/rating them would take a whole lot more tinkering and should be put off to a later time.

**Comment by Alys** [comment]
So the rating system would be for mods/staff only - I withdraw all objections. :)

I also have no objections to the list of users being done later (staged approaches are great) but I'm wondering if you could reuse the Hall of Heroes code to easily make a Hall of Villains, visible to only mods/staff. The "Reward User" fields could be copied to "Punish User" -- "Contrib Tier" becomes "Total Infraction Severity", "Contributions" becomes "Infractions".

Although possibly "Villains" and "Punish" are harsher terms than we should be using. :)

![screen shot 2014-12-24 at 12 11 23 pm](https://cloud.githubusercontent.com/assets/1495809/5545169/5a18a042-8b66-11e4-8842-2023a8888159.png)

**Comment by crookedneighbor** [comment]
Added a new item to the Discussion section:

> Automatic email sent to the person who wrote the offending message when post is hidden explaining that the community believed the post was in violation of the community guidelines, assuring the person that in all likelihood, their error was done in ignorance and encouraging them to review the community guidelines to avoid future posts from being hidden.

And moved this to Features:

> Allow reporters to add a message to the mods when reporting abuse

And the hall of villains items to a new section called Ideas for Part 3

**Comment by crookedneighbor** [comment]
I left this in Discussion:

> If a user has more than X flags on any of their posts in a certain time period, all future posts from them are automatically hidden.

Seemed like it would be trickier to implement than the other items, so I didn't want to move it to features just yet.

**Comment by crookedneighbor** [comment]
@Alys Are you sure the contributor notes are private?

I was checking on another issue, and used the /static/api page to check the [member details](https://habitrpg.com/static/api#!/members/members_uuid_GET), and was able to see the contributor notes for another user.

**Comment by Alys** [comment]
The contributor notes aren't meant to be private. Anyone can see the items that you've earned tiers for.

**Comment by crookedneighbor** [comment]
Ahh, I just misunderstood what you were saying earlier. My mistake.

**Comment by crookedneighbor** [comment]
@Alys Your comment in G+ gave me an idea.

If a regular user clicks the flag button, an optional text box appears for the user to give an explanation of why they are flagging it.

If a mod flags it, the optional text box, if filled, will post a message. This should make it easier to flag and then notify the user that their comment was flagged. It could send a pm instead, if preferred.

Alternatively, the optional text box could fill out an entry in the "hall of villains" automatically.

Anyway, those may be a half-baked ideas, so feel free to shoot them down.

**Comment by Alys** [comment]
When something needs to be censored, I typically want to hide it immediately. When I write the explanation messages, I sometimes need to take some time over it to choose my words carefully. I wouldn't use the flag dialog's text box for the message because I'd be too stressed over leaving the message visible while I wrote. :) 

Hall of Villians (semi-)automation would be very nice, but again I'd want to type any messages there after flagging.

**Comment by crookedneighbor** [comment]
@Alys I thought I responded to this, but apparently not. That's a good point. I take back my suggestion!

**Comment by librarianmage** [comment]
Is there anything that should be done first?

**Comment by Alys** [comment]
This still needs discussion. Some thoughts:

- [ ] _Add report dialog to block user button_ - I've logged an issue for flagging profiles: https://github.com/HabitRPG/habitica/issues/9486 . It's not tied to the block button but I think it's good to keep the two features separate (you might block someone you don't like who hasn't violated the rules; having a report button the same as on chat messages will make it more obvious that you can report users / profiles). I've edited the original post here to include that link.
- [x] _Allow reporters to add a message to the mods when reporting abuse_ - We've wanted that for a while and still do I believe.
- _If a user has more than X flags on any of their posts in a certain time period, all future posts from them are automatically hidden._ - While I like this idea for cases of repeated abuse, I don't think it's important enough to spend coding and reviewing time on it (at least not now while there's so many bugs). Mods are pretty good at remembering repeat offenders and we discuss muting them when we feel it might be appropriate. Also, this feature could be used maliciously to harass a player by flagging many of their posts, thereby muting them (mods would catch and fix that quickly but it would still be stressful for the attacked player). There are also rare cases where mods flag several messages at once but wouldn't want the user muted.
- _Automatic email sent to the person who wrote the offending message when post is hidden explaining that the community believed the post was in violation of the community guidelines, assuring the person that in all likelihood, their error was done in ignorance and encouraging them to review the community guidelines to avoid future posts from being hidden._  - This is no longer desired. There are rare cases where messages need to be hidden for sensitive reasons but we wouldn't want the user to be emailed about it (e.g., for under-aged players). It would also be distressing to an innocent player if their posts were flagged maliciously, which does happen occasionally.
- [ ] _List of reported users that only admins can see_ - Useful but not high priority I believe. Probably not worth creating an issue for this until the bug list is smaller. Database admins can extract the list now if essential.
- _Mods can assign comments (similar to contributor tier notes) and scores to infractions (1-5), after a user's total score reaches a certain threshold, chat privileges are revoked_ - In my opinion this would be nice but again perhaps not something we'd want a contributor to work on right now. EDIT April 2018: I'm no longer sure this is worth considering. We'd probably spend more time adding scores than we'd save from having chat permissions revoked automatically, and we'd probably want to revoke them manually anyway as a final human check on the situation.

**Comment by veeeeeee** [comment]
We should revisit this in light of the upcoming admin panel feature to determine which needs are still necessary and how to best roll this out.

**Comment by saraolson** [comment]
we've made quite a few updates to our reporting tools over the years as well as the content and ways players can report violations. the feature set as it exists now fits our needs, so going to close this issue.
